lnpbp-0021.md

```diff
@@ -1,213 +1,183 @@
-            format: StateFormat::Declarative,
-            abi: none!()
-        }
-    },
-    public_right_types: none!(),
-}
+- [Abstract](#abstract)
+- [Background](#background)
+- [Motivation](#motivation)
+- [Design](#design)
+- [Specification](#specification)
+- [Compatibility](#compatibility)
+- [Rationale](#rationale)
+- [Reference implementation](#reference-implementation)
+- [Acknowledgements](#acknowledgements)
+- [References](#references)
+- [Copyright](#copyright)
+- [Test vectors](#test-vectors)
+
+
+## Abstract
+
+
+## Background
+
+
+## Motivation
+
+
+## Design
+
+- [x] Media as URI (which can be attachments distributed with Storm or usual URLs)
+- [x] Small media ("previews" if given together with large media)
+- [x] Fraction of assets
+- [x] Engravings
+- [x] Reserves
+- [x] Possible decentralized issue
+
+
+## Specification
+
+Interface specification is the following Contractum code:
+
+```haskell
+-- # Defining main data structures
+
+-- collectibles are usually scarse, so we limit their max number to 64k
+data ItemsCount :: U32
+
+-- each collectible item is unique and must have an id
+data TokenId :: U16
+
+-- Collectibles are owned in fractions of a single item; here the owned
+-- fraction means `1/F`. Ownership of the full item is specified `F=1`;
+-- half of an item as `F=2` etc.
+data OwnedFraction :: U64
+
+data TxOut :: txid [Byte ^ 32], vout U16
+
+-- allocation of a single token or its fraction to some transaction output
+data Allocation :: TokenId, OwnedFraction
+
+data Media ::
+    type MimeType,
+    data [Byte]
+
+data Attachment ::
+    type MimeType,
+    uri: Uri
+
+data POR :: -- proof of reserves
+    reserves TxOut,
+    proof [Byte] -- schema-specific proof 
+
+data Token ::
+    id TokenId,
+    name [Ascii ^ 1..40],
+    details [Unicode ^ 40..256]?,
+    preview Media?, -- always embedded preview media < 64kb
+    media Attachment?, -- external media which is the main media for the token
+    attachments { U8 ^ ..20 -> Attachment } -- auxillary attachments by type (up to 20 attachments)
+    reserves POR? -- output containing locked bitcoins; how reserves are
+                  -- proved is a matter of a specific schema implementation
+
+data Denomination :: 
+    ticker [Ascii ^ 1..8],
+    name [Ascii ^ 1..40],
+    details [Unicode ^ 40..256]?,
+    contract [Unicode]??,
+
+data Auxiliary ::
+    -- list of allowed sources
+    sources [Source]
+    
+     -- each attachment type is a mapping from attachment id 
+     -- (used as `Token.attachments` keys) to a short Ascii string
+     -- verbally explaining the type of the attachment for the UI
+     -- (like "sample" etc).
+    attachmentTypes { U8 ^ ..32 -> [Ascii ^ 1..20] }
+
+data Source :: url(Proto, Dns) | urn(UrnPrefix) | storm(NodeAddr)
+data Proto :: http | https | httpxk | ws | wss | wssxk
+
+interface RGB21
+    global Name :: Denomination
+    global Tokens :: [(Token, [Engraving])]
```

Comment:
```suggestion
    global Tokens :: Token+
    global Engravings :: Engraving+
```

Each global state field in a contract operates either as a replacable single item - or as an append-only list of items - otherwise it is impossible to have a consensus  in client-side-validation. Thus, we introduce the same notation as with the owned state: `+` suffix for the state which can be defined multiple times; `?` for optional state, `*` for state which may be absent or present as a single item (and once defined can be never removed) or absence of suffix for the state which is always present and only can be replaced.

As for the engravings, due to this fact we need to separate them from the token definitions.